### PR TITLE
[SPARK-39957][CORE] Delay onDisconnected to enable Driver receives ExecutorExitCode

### PR DESCRIPTION
**Pull request by kevin85421**
### What changes were proposed in this pull request?
When onDisconnected is triggered,

(1) Delay `RemoveExecutor` for 5 seconds to enable driver receives ExecutorExitCode from slow path
(2) Prevent task scheduler from assigning tasks on the lost executor. (By adding the executor to `executorsPendingLossReason`)

### Why are the changes needed?
There are two methods to detect executor loss.

(1) (fast path) `onDisconnected` Executor -> Driver:
When Executor closes its JVM, the socket (Netty's channel) will be closed. The function onDisconnected will be triggered when it knows the channel is closed.

(2) (slow path) ExecutorRunner -> Worker -> Master -> Driver (See #37385  for details)
When executor exits with ExecutorExitCode, the exit code will be passed from ExecutorRunner to Driver.

Because fast path determines the executor loss without the information of ExecutorExitCode, these two methods may categorize same cases into different conclusions. For example, when Executor exits with ExecutorExitCode HEARTBEAT_FAILURE, onDisconnected will consider the executor loss as a task failure, but slow path will consider it as a network failure. Obviously, HEARTBEAT_FAILURE is a network failure.

[Notice]
For more details about ExecutorExitCode, check #37385 for more details.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
```bash
build/sbt "core/testOnly *SparkContextSuite -- -z ExitCode"
```

### CODE REVIEW

**Comment by kevin85421** [comment]
cc. @Ngone51 @jiangxb1987

File: core/src/main/scala/org/apache/spark/scheduler/cluster/StandaloneSchedulerBackend.scala [hunk 1]
```diff
@@ -175,6 +182,11 @@ private[spark] class StandaloneSchedulerBackend(
       exitStatus: Option[Int],
       workerHost: Option[String]): Unit = {
     val reason: ExecutorLossReason = exitStatus match {
+      case Some(ExecutorExitCode.HEARTBEAT_FAILURE) =>
+        ExecutorExited(ExecutorExitCode.HEARTBEAT_FAILURE, exitCausedByApp = false, message)
+      case Some(ExecutorExitCode.DISK_STORE_FAILED_TO_CREATE_DIR) =>
+        ExecutorExited(ExecutorExitCode.DISK_STORE_FAILED_TO_CREATE_DIR,
+          exitCausedByApp = false, message)
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/StandaloneSchedulerBackend.scala:189]: Could you also include the case where `workerHost` is non-empty? In that case, executor loss is caused by worker loss, so `exitCausedByApp` should be `false` too.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 2]
```diff
@@ -2398,4 +2398,13 @@ package object config {
       .version("3.3.0")
       .intConf
       .createWithDefault(5)
+
+  private[spark] val EXECUTOR_REMOVE_DELAY =
+    ConfigBuilder("spark.executor.removeDelayOnDisconnection")
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:2403]: This is specific to standalone right ? Or is it relevant to other resource managers ? (would impact the naming of the config)

Comment (kevin85421) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala:2403]: Yes, this PR only focuses on standalone. Do you have any suggestion about the config name? Maybe `spark.standalone.removeExecutorDelay`? Thank you!

File: core/src/main/scala/org/apache/spark/scheduler/cluster/StandaloneSchedulerBackend.scala (hunk 1 above)
Comment (kevin85421) [reply core/src/main/scala/org/apache/spark/scheduler/cluster/StandaloneSchedulerBackend.scala:189]: Good point. Updated https://github.com/apache/spark/pull/37400/commits/55080195dc6326d3fa3886da83dd0cbaec04a26c.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala (hunk 2 above)
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala:2403]: How about `spark.standalone.executorRemoveDelayOnDisconnection`?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/StandaloneSchedulerBackend.scala [hunk 3]
```diff
@@ -175,8 +182,13 @@ private[spark] class StandaloneSchedulerBackend(
...
       case Some(code) => ExecutorExited(code, exitCausedByApp = true, message)
-      case None => ExecutorProcessLost(message, workerHost)
+      case None => ExecutorProcessLost(message, workerHost, causedByApp = workerHost == None)
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/StandaloneSchedulerBackend.scala:191]:
```suggestion
      case None => ExecutorProcessLost(message, workerHost, causedByApp = workerHost.isEmpty)
```

File: core/src/test/scala/org/apache/spark/SparkContextSuite.scala [hunk 4]
```diff
@@ -1354,6 +1355,50 @@ class SparkContextSuite extends SparkFunSuite with LocalSparkContext with Eventu
     }.getMessage
     assert(msg.contains("Cannot use the keyword 'proxy' or 'history' in reverse proxy URL"))
   }
+
+  test("ExitCode HEARTBEAT_FAILURE should be counted as network failure") {
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/SparkContextSuite.scala:1359]:
```suggestion
  test("SPARK-39557: ExitCode HEARTBEAT_FAILURE should be counted as network failure") {
```

File: core/src/test/scala/org/apache/spark/SparkContextSuite.scala [hunk 5]
```diff
@@ -1354,6 +1355,50 @@ class SparkContextSuite extends SparkFunSuite with LocalSparkContext with Eventu
...
+        System.exit(ExecutorExitCode.HEARTBEAT_FAILURE)
+      } else {
+        x
+      }
+    }.count()
+    assert(result == 10L)
+    sc.stop()
+  }
+
+  test("ExitCode HEARTBEAT_FAILURE will be counted as task failure when EXECUTOR_REMOVE_DELAY is" +
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/SparkContextSuite.scala:1380]:
```suggestion
  test("SPARK-39557: ExitCode HEARTBEAT_FAILURE will be counted as task failure when EXECUTOR_REMOVE_DELAY is" +
```

Comment (kevin85421) [reply core/src/test/scala/org/apache/spark/SparkContextSuite.scala]: Updated https://github.com/apache/spark/pull/37400/commits/80b86910a59f5b409caa17d46305027ed89c3dd1

File: core/src/main/scala/org/apache/spark/internal/config/package.scala (hunk 2 above)
Comment (kevin85421) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala]: Updated.

**Comment by kevin85421** [comment]
Gentle ping @Ngone51 @mridulm

**Comment by mridulm** [comment]
I will let @Ngone51 review/merge - I do not have a lot of context on standalone :-)

**Review by Ngone51 — APPROVED** [review]

**Comment by Ngone51** [comment]
Thanks @kevin85421 @mridulm , merged to Master!

**Comment by kevin85421** [comment]
Thank @mridulm @Ngone51 for the review!

File: core/src/main/scala/org/apache/spark/scheduler/cluster/StandaloneSchedulerBackend.scala [hunk 6]
```diff
@@ -272,4 +285,66 @@ private[spark] class StandaloneSchedulerBackend(
     }
   }
 
+  override def createDriverEndpoint(): DriverEndpoint = {
+    new StandaloneDriverEndpoint()
+  }
+
+  private class StandaloneDriverEndpoint extends DriverEndpoint {
+    // [SC-104659]: There are two paths to detect executor loss.
```
Comment (LuciferYang) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/StandaloneSchedulerBackend.scala:293]: @kevin85421 Sorry to bother you, but can you explain what `SC-104659` and `SC-104335` mean? This seems doesn't Spark jira. Thanks ~

**Pull request closed without merging**